File: src/org/usfirst/frc/team4322/command/Command.java
package org.usfirst.frc.team4322.command;
import java.util.ArrayList;
import java.util.function.*;

/**
 * Created by nicolasmachado on 4/20/16.
 */
public abstract class Command implements Runnable
{

	private boolean started = false;
	private boolean hasRun = false;
	private long startTime;
	private long timeout = 0;
	private ArrayList<Subsystem> subsystems = new ArrayList<>();

	public Command(ArrayList<Subsystem> subsystems, long timeout)
	{
		this.timeout = timeout;
		this.subsystems = subsystems;
	}

	public Command()
	{

	}

	protected void setTimeout(long millis)
	{
		this.timeout = millis;
	}

	public ArrayList<Subsystem> getSubsystems()
	{
		return subsystems;
	}

    protected abstract void initialize();

	public void require(Subsystem s)
	{
		subsystems.add(s);
	}
	public void require(Subsystem... s)
	{
		for(Subsystem sys : s)
		{
			subsystems.add(sys);
		}
	}

	public void cancel()
	{
		end();
		Scheduler.getInstance().remove(this);
	}

    protected abstract void end();

    protected abstract void interrupted();

	public void start()
	{
		startTime = System.currentTimeMillis();
		started = true;
		initialize();
		Scheduler.getInstance().add(this);
	}
	public long runTimeMillis()
	{
		return System.currentTimeMillis() - startTime;
	}

	protected abstract boolean isFinished();

	public boolean shouldRun()
	{
		if(timeout > 0)
		{
			return !isFinished() && timeout < runTimeMillis();
		}
		return !isFinished();
	}

    protected abstract void execute();

	public void run()
	{
		if(shouldRun())
		{
			execute();
			if(!hasRun)
			{
				hasRun = true;
			}
		}
		else
		{
			//All commands must run once.
			if(!hasRun)
			{
				execute();
				hasRun = true;
			}
			cancel();
		}

	}
	void interrupt()
	{
		interrupted();
	}

	public boolean isStarted()
	{
		return started;
	}

	public long getStartTime()
	{
		return startTime;
	}
}

File: src/org/usfirst/frc/team4322/command/Command.kt
package org.usfirst.frc.team4322.command

import java.util.*
import java.util.concurrent.ScheduledFuture

/**
 * Created by nicolasmachado on 4/20/16.
 */
abstract class Command : Runnable {

    var isStarted = false
        private set
    private lateinit var future : ScheduledFuture<*>
    var hasRun = false
    var isDone = false
    var startTime: Long = 0
        private set
    private var timeout: Long = 0
    val subsystems : ArrayList<Subsystem>

    protected abstract val isFinished: Boolean

    constructor(subsystems: ArrayList<Subsystem>, timeout: Long) {
        this.timeout = timeout
        this.subsystems = subsystems
    }

    constructor() {
       subsystems = ArrayList()
    }

    protected fun setTimeout(millis: Long) {
        this.timeout = millis
    }

    protected abstract fun initialize()

    fun require(s: Subsystem) {
        subsystems.add(s)
    }

    fun require(vararg s: Subsystem) {
        subsystems += s
    }

    fun cancel() {
        if(!isDone) {
            end()
            isStarted = false
            isDone = true
            future.cancel(true)
        }
    }

    protected abstract fun end()

    protected abstract fun interrupted()

    open fun start() {
        if(!isStarted) {
            initialize()
            future = Scheduler.add(this)
            startTime = System.currentTimeMillis()
            isStarted = true
            hasRun = false
            isDone = false
        }
    }

    fun runTimeMillis(): Long {
        return System.currentTimeMillis() - startTime
    }

    fun shouldRun(): Boolean {
        return if (timeout > 0) {
            !isFinished && timeout < runTimeMillis()
        } else !isFinished
    }

    protected abstract fun execute()

    override fun run() {
        if (shouldRun()) {
            execute()
            if (!hasRun) {
                hasRun = true
            }
        } else {
            //All commands must run once.
            if (!hasRun) {
                execute()
                hasRun = true
            }
            cancel()
        }

    }

    internal fun interrupt() {
        interrupted()
    }
}
